[QoI] Improve diagnostic for raw representable argument mismatch

xedin opened the pull request:
Covers some common situations when switching from literal types e.g.
string or int to more Swift Way of raw representable enum of choices.

Resolves: rdar://problem/32431165, rdar://problem/31977537, rdar://problem/32432253
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[QoI] Improve diagnostic for raw representable argument mismatch

Covers some common situations when switching from literal types e.g.
string or int to more Swift Way of raw representable enum of choices.

Resolves: rdar://problem/32431165, rdar://problem/31977537, rdar://problem/32432253

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

rudkx commented on test/Sema/enum_raw_representable.swift:
```diff
@@ -103,3 +103,40 @@ func rdar32431736() {
+// expected-error@-1 {{cannot convert value of type 'String' to expected argument type 'E_32431165'}} {{15-15=E_32431165(rawValue: }} {{19-19=)}}
+rdar32431165_1(42, "")
+// expected-error@-1 {{cannot convert value of type 'String' to expected argument type 'E_32431165'}} {{20-20=E_32431165(rawValue: }} {{22-22=)}}
+
```
Can you add a test for a nonexistent enum element?
```
_ = rdar32431165_1(.nada)
```

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
+      return false;
+  }
+
+  auto argType = argExpr->getType();
```
Can you use `ConstraintSystem::getType(argExpr)` here?

I'm inching closer to getting all the CS code switched over to the side map but new appearances of `Expr::getType()` keep creeping in.

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
+      KnownProtocolKind::ExpressibleByIntegerLiteral};
+
+  const auto *CS = CCI.CS;
+  auto arguments = decomposeArgType(argExpr->getType(), argLabels);
```
Same re: `getType()`.

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5317,6 +5317,96 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
+      for (auto kind : rawRepresentableProtocols) {
+        // If trying to convert from raw type to raw representable,
+        // or vice versa from raw representable (e.g. enum) to raw type.
+        if (isRawRepresentableMismatch(argType, paramType, kind, CS)) {
```
Can you change this to:
```
  if (!isRaw...)
    continue;
```

to reduce indentation?

rudkx reviewed (commented):
LGTM other than the handful of comments.

xedin commented:
Changed the `if` statement and added couple more tests as requested. Also talked to @rudkx offline and we've decided not to change `argExpr->getType` because there are couple of issues with that related to the expressions which are checked separately, he is going to take a look at the later.

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
